debian/simple-pwgen.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" (C) Copyright 2022 Himadri Sekhar Basu <hsb10@iitbbs.ac.in>,
.\"
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH simple-pwgen SECTION "May 29 2022"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
simple-pwgen \- program to do something
.SH SYNOPSIS
.B simple-pwgen
.RI [ options ] " files" ...
.br
.B bar
.RI [ options ] " files" ...
.SH DESCRIPTION
This manual page documents briefly the
.B simple-pwgen
and
.B bar
commands.
.PP
.\" TeX users may be more comfortable with the \fB<whatever>\fP and
.\" \fI<whatever>\fP escape sequences to invode bold face and italics,
.\" respectively.
\fBsimple-pwgen\fP is a program that...
.SH OPTIONS
These programs follow the usual GNU command line syntax, with long
options starting with two dashes (`-').
A summary of options is included below.
For a complete description, see the Info files.
.TP
.B \-h, \-\-help
Show summary of options.
.TP
.B \-v, \-\-version
Show version of program.
.SH SEE ALSO
.BR bar (1),
.BR baz (1).
.br
The programs are documented fully by
.IR "The Rise and Fall of a Fooish Bar" ,
available via the Info system.


debian/changelog
simple-pwgen (1.1.1) stable; urgency=medium

  * fix permissions of github actions
  * Use new gschema name (#49)
  * Update readme with organization mamolinux (#50)
  * Prepare to support other generation methods (#51)
    - Plan to include method for PIN generation
    - Plan to include method for password generation
      using diceware method
  * Encrypted string for better security; closes #17 (#52)
    - Generated password string is encrypted before
      any communication and decrypted just before use.
  * Update debian packaging (#53)
    - d/control: update compatibility to 13
    - d/rules: clean up unused rules and
      avoid inclusion of compiled gschema file
  * Add and update ui for methods (#54)
    - Add ui for pin generation and diceware method
    - Update copyright in ui files
  * Use image icon for show/hide and copy (#55)
    - Use password reveal icon to show or hide password
    - Use copy icon button to copy the password to clipboard

 -- Himadri Sekhar Basu <hsb10@iitbbs.ac.in>  Sun, 21 May 2023 09:00:36 +0530

simple-pwgen (1.1.0) stable; urgency=medium

  * fix publish-deb workflow (#38)
  * Add workflow to update mamolinux/stable branch
  * update badges url (#40)
  * Create snap of simple-pwgen (#41)
    - Add snap/snapcraft.yaml
  * Use relative path for ui and css files (#42)
    - Place UI and css files into source tree
    - Use relative path for UI files from source tree
  * Fix snap source and remove gui part (#43)
  * Remove python gui script (#44)
    - Add argument "-g" to run gui
    - Modify the desktop files
    - remove gui script from d/simple-pwgen.install
    - Remove unnecessary codes and white spaces
  * Precompile gschema (#45)
    - Precompile gschema files
    - Remove unnecessary codes
  * Change snap grade and confinement (#46)
  * Update readme from version 1.0.8 to 1.0.9 (#47)
  * Fix dbus access denied error for snap (#48)

 -- Himadri Sekhar Basu <hsb10@iitbbs.ac.in>  Wed, 15 Feb 2023 21:01:29 +0530

simple-pwgen (1.0.9) stable; urgency=medium

  * fix auto build and publish deb packaging workflow (#33)
  * Modify sources to accomodate cli interface (#34)
    - setup.cfg: update console and gui scripts
    - remove VERSION from source
    - update test file
    - src/init: remove all codes
    - src/common: append version and UI path
    - src/gui: append from common and remove cli function
    - src: create main function script
  * Update debian packaging and workflow (#35)
    - workflow: clean up unused code
    - remove dh_make example files
    - d/control: update runtime depends
    - d/python3-sp.install: rename cli script
    - d/README: update
    - rename salsa-ci
  * Update readme from version 1.0.7 to 1.0.8 (#36)
    - fix logo
    - Add dependency
    - Update assets link
  * Update installation instructions (#37)
  * Add workflow to update mamolinux/stable branch

 -- Himadri Sekhar Basu <hsb10@iitbbs.ac.in>  Tue, 28 Jun 2022 21:47:28 +0530

simple-pwgen (1.0.8) stable; urgency=medium

  * New icon for simple-pwgen
  * fix error "Permission denied"; fixes #25
  * update readme to release 1.0.7 (#26)
  * Update debian standards version
  * Update readme with new features
    - Add password strength and comments to features
    - Add new screenshots
  * fix typo in README
  * add new and update old workflows
    - Update workflows to run on all branches
    - add new workflow to publish deb package from latest release
    - Fix typo in README
    - Update test script
  * pythonize simple-pwgen (#27)
    - use pure python packaging methods
    - use debian revision number in d/changelog
    - update .gitignore
    - update desktop files
    - Add separate VERSION file to source
  * re-debianize to match python packaging (#28)
    - Create python-simple-pwgen package as commadline backend
    - Modify simple-pwgen to provide GUI of app
    - Create sphinx based documentation
  * Update CI workflow (#30)
  * fix launchpad build failures (#29)
    - d/control: reduce compat level and update depends
    - d/python3-simple-pwgen.install: remove installation of manual
    - d/rules: remove sphinx and update rules
  * fix CI workflow (#31)
  * fix versioning to comply with PEP 440 versioning (#32)

 -- Himadri Sekhar Basu <hsb10@iitbbs.ac.in>  Wed, 01 Jun 2022 22:42:26 +0530

simple-pwgen (1.0.7) stable; urgency=medium

  * Update copyright in About dialog
  * Update UI with strengthbar
    - Update strengthbar
    - Add comment box for password strength
    - Add entries for password score, entropy
    - Add entries for time and number of guesses to crack
    - Add entry for detailed comment on password
  * Update shortcut keys
    - Update tooltips on buttons
    - Add new shortcut keys for settings; closes #13
    - Update shortcut keys method
  * Show info on password
    - Show password strengthbar
    - Calculate password entropy, score etc.
  * Colorize Strengthbar; closes #10
    - Use color different colors for bar and comment
    - Update UI

 -- Himadri Sekhar Basu <hsb10@iitbbs.ac.in>  Mon, 07 Feb 2022 12:13:44 +0530

simple-pwgen (1.0.6) stable; urgency=medium

  * Add keyboard shortcuts
    
    - Show keyboard shortcuts in menus
    - Crtl+G to generate
    - Crtl+H to Show/Hide
    - Crtl+C to copy to clipboard
  * Add gitignore
  * Update README and remove snapcraft.yaml

 -- Himadri Sekhar Basu <hsb10@iitbbs.ac.in>  Wed, 29 Dec 2021 20:13:56 +0530

simple-pwgen (1.0.5-0patch1) stable; urgency=medium

  * quick fix exclude symbol and generated password entry

 -- Himadri Sekhar Basu <hsb10@iitbbs.ac.in>  Tue, 09 Nov 2021 09:50:02 +0530

simple-pwgen (1.0.5) stable; urgency=medium

  * Update snap support
  * Update debian compatibility 9 to 10
  * Hide generated password to avoid shoulder-surfer.; fixes #19
  * Add snap support (#18)
  * Update svg icon to fix Gdk-pixbuff error. (#16)

 -- Himadri Sekhar Basu <hsb10@iitbbs.ac.in>  Mon, 08 Nov 2021 12:22:47 +0530

simple-pwgen (1.0.4) stable; urgency=medium


usr/lib/simple-pwgen/simple-pwgen.py
#!/usr/bin/python3

# Copyright (C) 2021 Himadri Sekhar Basu <hsb10@iitbbs.ac.in>
#
# This file is part of simple-pwgen.
#
# simple-pwgen is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# simple-pwgen is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with simple-pwgen. If not, see <http://www.gnu.org/licenses/>
# or write to the Free Software Foundation, Inc., 51 Franklin Street,
# Fifth Floor, Boston, MA 02110-1301, USA..
#
# Author: Himadri Sekhar Basu <hsb10@iitbbs.ac.in>
#

# import the necessary modules!
import gettext
import gi
import locale
import setproctitle
import warnings

# Suppress GTK deprecation warnings
warnings.filterwarnings("ignore")

gi.require_version("Gtk", "3.0")
from gi.repository import Gtk, Gio, Gdk

from common import CONFIG_FILE, PasswordGenerator

setproctitle.setproctitle("simple-pwgen")

# i18n
APP = 'simple-pwgen'
LOCALE_DIR = "/usr/share/locale"
locale.bindtextdomain(APP, LOCALE_DIR)
gettext.bindtextdomain(APP, LOCALE_DIR)
gettext.textdomain(APP)
_ = gettext.gettext

class simple_pwgen(Gtk.Application):
	# Main initialization routine
	def __init__(self, application_id, flags):
		Gtk.Application.__init__(self, application_id=application_id, flags=flags)
		self.connect("activate", self.activate)

	def activate(self, application):
		windows = self.get_windows()
		if (len(windows) > 0):
			window = windows[0]
			window.present()
			window.show()
		else:
			window = SimplepwgenWindow(self)
			self.add_window(window.window)
			window.window.show()
	
class SimplepwgenWindow():
	
	def __init__(self, application):
		
		self.application = application
		self.settings = Gio.Settings(schema_id="org.x.simple-pwgen")
		self.generator = PasswordGenerator()
		self.icon_theme = Gtk.IconTheme.get_default()
		
		# Set the Glade file
		gladefile = "/usr/share/simple-pwgen/simple-pwgen.ui"
		self.builder = Gtk.Builder()
		self.builder.add_from_file(gladefile)
		self.window = self.builder.get_object("MainWindow")
		self.window.set_title(_("Simple Password Generator"))
		
		# Create variables to quickly access dynamic widgets
		# input values
		self.pwlengthfield = self.builder.get_object("pwlengthfield")
		self.lcase_switch = self.builder.get_object("lcase_switch")
		self.minlcase_num = self.builder.get_object("minlcase")
		self.excludelcases_field = self.builder.get_object("excludelcases")
		self.ucase_switch = self.builder.get_object("ucase_switch")
		self.minUcase_num = self.builder.get_object("minucase")
		self.excludeUcases_field = self.builder.get_object("excludeucases")
		self.digit_switch = self.builder.get_object("digit_switch")
		self.mindigits_num = self.builder.get_object("mindigits")
		self.excludedigits_field = self.builder.get_object("excludedigits")
		self.symbol_switch = self.builder.get_object("symbol_switch")
		self.minsymbol_num = self.builder.get_object("minsymbol")
		self.excludesymbol_field = self.builder.get_object("excludesymbols")
		
		# output values
		self.passwordfield = self.builder.get_object("passwordfield")
		self.strength_bar = self.builder.get_object("strength_bar")
		self.pw_strength_field = self.builder.get_object("pw_strength_field")
		self.pw_score_field = self.builder.get_object("pw_score_field")
		self.pw_guesses_field = self.builder.get_object("pw_guesses_field")
		self.pw_entropy_field = self.builder.get_object("pw_entropy_field")
		self.pw_crack_time_field = self.builder.get_object("pw_crack_time_field")
		self.pw_comment_field = self.builder.get_object("pw_comment_field")
		
		# Buttons
		self.reset_button = self.builder.get_object("reset_button")
		self.save_button = self.builder.get_object("save_button")
		self.generate_button = self.builder.get_object("generate_button")
		self.showhide_button = self.builder.get_object("showhide_button")
		self.copy_button = self.builder.get_object("copy_button")
		self.quit_button = self.builder.get_object("quit_button")
		
		# Widget signals
		self.reset_button.connect("clicked", self.on_reset_button)
		self.save_button.connect("clicked", self.on_save_button)
		self.generate_button.connect("clicked", self.on_generate_button)
		self.showhide_button.connect("clicked", self.on_showhide_button)
		self.copy_button.connect("clicked", self.on_copy_button)
		self.quit_button.connect("clicked", self.on_quit)
		self.passwordfield.connect("changed", self.show_pwstrength)
		
		# Menubar
		accel_group = Gtk.AccelGroup()
		self.window.add_accel_group(accel_group)
		menu = self.builder.get_object("main_menu")
		# Add "Shortcuts" option in drop-down menu
		item = Gtk.ImageMenuItem()
		item.set_image(Gtk.Image.new_from_icon_name("preferences-desktop-keyboard-shortcuts-symbolic", Gtk.IconSize.MENU))
		item.set_label(_("Keyboard Shortcuts"))
		item.connect("activate", self.open_keyboard_shortcuts)
		key, mod = Gtk.accelerator_parse("<Control>K")
		item.add_accelerator("activate", accel_group, key, mod, Gtk.AccelFlags.VISIBLE)
		menu.append(item)
		# Add "About" option in drop-down menu
		item = Gtk.ImageMenuItem()
		item.set_image(Gtk.Image.new_from_icon_name("help-about-symbolic", Gtk.IconSize.MENU))
		item.set_label(_("About"))
		item.connect("activate", self.open_about)
		key, mod = Gtk.accelerator_parse("<Control>F1")
		item.add_accelerator("activate", accel_group, key, mod, Gtk.AccelFlags.VISIBLE)
		menu.append(item)
		# Add "Quit" option in drop-down menu
		item = Gtk.ImageMenuItem(label=_("Quit"))
		image = Gtk.Image.new_from_icon_name("application-exit-symbolic", Gtk.IconSize.MENU)
		item.set_image(image)
		item.connect('activate', self.on_quit)
		key, mod = Gtk.accelerator_parse("<Control>Q")
		item.add_accelerator("activate", accel_group, key, mod, Gtk.AccelFlags.VISIBLE)
		key, mod = Gtk.accelerator_parse("<Control>W")
		item.add_accelerator("activate", accel_group, key, mod, Gtk.AccelFlags.VISIBLE)
		menu.append(item)
		# Show all drop-down menu options
		menu.show_all()
		
		self.load_conf()
		self.load_css(color="inherit")
		
	def load_conf(self):
		
		self.generator.load_config()
		self.pwlengthfield.set_text(str(self.generator.pwlength))
		
		self.lcase_switch.set_active(self.generator.lowercase)
		self.minlcase_num.set_text(str(self.generator.lowercase_num))
		self.excludelcases_field.set_text(str(self.generator.excludeLowercase))
		if not self.generator.lowercase:
			self.minlcase_num.set_sensitive(False)
			self.excludelcases_field.set_sensitive(False)
		else:
			self.minlcase_num.set_sensitive(True)
			self.excludelcases_field.set_sensitive(True)
		
		self.ucase_switch.set_active(self.generator.uppercase)
		self.minUcase_num.set_text(str(self.generator.uppercase_num))
		self.excludeUcases_field.set_text(str(self.generator.excludeUppercase))
		if not self.generator.uppercase:
			self.minUcase_num.set_sensitive(False)
			self.excludeUcases_field.set_sensitive(False)
		else:
			self.minUcase_num.set_sensitive(True)
			self.excludeUcases_field.set_sensitive(True)
		
		self.digit_switch.set_active(self.generator.digit)
		self.mindigits_num.set_text(str(self.generator.digit_num))
		self.excludedigits_field.set_text(str(self.generator.excludeDigit))
		if not self.generator.digit:
			self.mindigits_num.set_sensitive(False)
			self.excludedigits_field.set_sensitive(False)
		else:
			self.mindigits_num.set_sensitive(True)
			self.excludedigits_field.set_sensitive(True)
			
		self.symbol_switch.set_active(self.generator.symbol)
		self.minsymbol_num.set_text(str(self.generator.symbol_num))
		self.excludesymbol_field.set_text(str(self.generator.excludeSymbol))
		if not self.generator.symbol:
			self.minsymbol_num.set_sensitive(False)
			self.excludesymbol_field.set_sensitive(False)
		else:
			self.minsymbol_num.set_sensitive(True)
			self.excludesymbol_field.set_sensitive(True)
	
	def open_keyboard_shortcuts(self, widget):
		gladefile = "/usr/share/simple-pwgen/shortcuts.ui"
		builder = Gtk.Builder()
		builder.set_translation_domain(APP)
		builder.add_from_file(gladefile)
		window = builder.get_object("shortcuts-simplepwgen")
		window.set_title(_("Simple Password Generator"))
		window.show()
	
	def open_about(self, widget):
		dlg = Gtk.AboutDialog()
		dlg.set_transient_for(self.window)
		dlg.set_icon_name("simple-pwgen")
		dlg.set_logo_icon_name("simple-pwgen")
		dlg.set_title(_("About"))
		
		dlg.set_program_name(_("Simple Password Generator"))
		dlg.set_version("__DEB_VERSION__")
		dlg.set_comments(_("Very simple Python3-based GUI application to generate secure and random password."))
		dlg.set_website("https://hsbasu.github.io/simple-pwgen")
		dlg.set_copyright("Copyright \xa9 2022 Himadri Sekhar Basu")
		dlg.set_authors(["Himadri Sekhar Basu <https://hsbasu.github.io>"])
		dlg.set_documenters(["Himadri Sekhar Basu <https://hsbasu.github.io>"])
		try:
			h = open('/usr/share/common-licenses/GPL', encoding="utf-8")
			s = h.readlines()
			gpl = ""
			for line in s:
				gpl += line
			h.close()
			dlg.set_license(gpl)
		except Exception as e:
			print (e)
		
		def close(w, res):
			if res == Gtk.ResponseType.CANCEL or res == Gtk.ResponseType.DELETE_EVENT:
				w.destroy()
		dlg.connect("response", close)
		dlg.show()
	
	def on_quit(self, widget):
		self.application.quit()
	
	def on_reset_button(self, widget):
		try:
			self.config.read(CONFIG_FILE)
			self.pwlengthfield.set_text(self.config["default"]['pwlength'])
			self.lcase_switch.set_active(int(self.config["default"]['lowercase']))
			self.minlcase_num.set_text(self.config["default"]['lowercase_num'])
			self.excludelcases_field.set_text(self.config["default"]['excludeLowercase'])
			self.ucase_switch.set_active(int(self.config["default"]['uppercase']))
			self.minUcase_num.set_text(self.config["default"]['uppercase_num'])
			self.excludeUcases_field.set_text(self.config["default"]['excludeUppercase'])
			self.digit_switch.set_active(int(self.config["default"]['digit']))
			self.mindigits_num.set_text(self.config["default"]['digit_num'])
			self.excludedigits_field.set_text(self.config["default"]['excludeDigit'])
			self.symbol_switch.set_active(int(self.config["default"]['symbol']))
			self.minsymbol_num.set_text(self.config["default"]['symbol_num'])
			self.excludesymbol_field.set_text(self.config["default"]['excludeSymbol'])
		except:
			self.pwlengthfield.set_text("8")
			self.lcase_switch.set_active(1)
			self.minlcase_num.set_text("1")
			self.excludelcases_field.set_text("")
			self.ucase_switch.set_active(1)
			self.minUcase_num.set_text("1")
			self.excludeUcases_field.set_text("")
			self.digit_switch.set_active(1)
			self.mindigits_num.set_text("1")
			self.excludedigits_field.set_text("")
			self.symbol_switch.set_active(0)
			self.minsymbol_num.set_text("0")
			self.excludesymbol_field.set_text("")
	
	def on_save_button(self, widget):
		"""Saves user configurations to config file.

		Saves user-defined configurations to config file.
		If the config file does not exist, it creates a new
		config file (~/.config/simple-pwgen/config.cfg)
		in user's home directory.
		"""
		if self.lcase_switch.get_active():
			lcase = 1
		else:
			lcase = 0
		if self.ucase_switch.get_active():
			ucase = 1
		else:
			ucase = 0
		if self.digit_switch.get_active():
			digit = 1
		else:
			digit = 0
		if self.symbol_switch.get_active():
			symbol = 1
		else:
			symbol = 0
			
		self.generator.config['user'] = {
			'pwlength': self.pwlengthfield.get_text(),
			'lowercase': lcase,
			'lowercase_num': self.minlcase_num.get_text(),
			'excludeLowercase': self.excludelcases_field.get_text(),
			'uppercase': ucase,
			'uppercase_num': self.minUcase_num.get_text(),
			'excludeUppercase': self.excludeUcases_field.get_text(),
			'digit': digit,
			'digit_num': self.mindigits_num.get_text(),
			'excludeDigit': self.excludedigits_field.get_text(),
			'symbol': symbol,
			'symbol_num': self.minsymbol_num.get_text(),
			'excludeSymbol': self.excludesymbol_field.get_text()
		}
		with open(CONFIG_FILE, 'w') as f:
				self.generator.config.write(f)
		
		self.load_conf()
	
	def on_generate_button(self, widget):
		self.password = self.generator.GeneratePW()
		self.passwordfield.set_text(self.password)
	
	def on_showhide_button(self, widget):
		if self.passwordfield.get_visibility() == False:
			self.passwordfield.set_visibility(True)
			self.showhide_button.set_label("Hide")
		else:
			self.passwordfield.set_visibility(False)
			self.showhide_button.set_label("Show")
	
	def on_copy_button(self, widget):
		self.clipboard = Gtk.Clipboard.get(Gdk.SELECTION_CLIPBOARD)
		self.clipboard.set_text(self.password, -1)
		
	def load_css(self, color: str):
		css_provider = Gtk.CssProvider()
		Priority = Gtk.STYLE_PROVIDER_PRIORITY_USER
		css = open('/usr/share/simple-pwgen/style.css', 'r')
		css_rawdata = css.read() % (color, color)
		# print(css_rawdata)
		css_data = bytes(css_rawdata, 'utf-8')
		css.close()
		css_provider.load_from_data(css_data)
		Gtk.StyleContext.add_provider_for_screen(Gdk.Screen.get_default(), css_provider, Priority)
		
	def show_pwstrength(self, widget):
		password = self.passwordfield.get_text()
		[pw_score, pw_comment, color] = self.generator.check_pwstrength(password)
		[pw_strength, pw_entropy, num_guess_crack, timerq_crack] = self.generator.check_pwentrpy(password)
		
		self.load_css(color)
		
		self.strength_bar.set_value(pw_score/10)
		self.pw_strength_field.set_text(pw_strength)
		self.pw_score_field.set_text(str(pw_score))
		self.pw_guesses_field.set_text(str(num_guess_crack))
		self.pw_entropy_field.set_text(str(pw_entropy))
		self.pw_crack_time_field.set_text(str(timerq_crack))
		self.pw_comment_field.set_text(pw_comment)
		

if __name__ == "__main__":
	application = simple_pwgen("org.x.simple-pwgen", Gio.ApplicationFlags.FLAGS_NONE)
	application.run()
	